### PR TITLE
feat: implemented a param to specify any vaulted payment method type that they want to block.

### PR DESCRIPTION
**Pull request by ale-grosselle**
### Summary

~~Implemented `vaultedPaypalAccountsDisabled` argument.  
If set to true, it force to disable paypal vault.~~

Implemented `vaultedPaymentMethodTypesThatShouldBeHidden`. 
It accepts an array of paymentMethodTypes (e.g CreditCard, PayPalAccount) and specify any vaulted payment method type that want to block. Important: ApplePayCard, VenmoAccount, AndroidPayCard are always blocked.

This is very useful when you have two different payment flows and on one it is not possible to enable paypal vaulting (due to legal restriction for example).

### Checklist

- [ ] Added a changelog entry

-

### CODE REVIEW

**Comment by crookedneighbor** [comment]
I think this should be done more holistically and allow the merchant to specify any vaulted payment method type that they want to block. 

Also, I think it should be a parameter under `vaultManager`, not a top level parameter.

**Comment by crookedneighbor** [comment]
I think this should be done more holistically and allow the merchant to specify any vaulted payment method type that they want to block. 

Also, I think it should be a parameter under `vaultManager`, not a top level parameter.

**Comment by ale-grosselle** [comment]
@crookedneighbor good point, I'm going to implement your suggestions!

**Comment by ale-grosselle** [comment]
@crookedneighbor "_I think it should be a parameter under vaultManager_", could you explain to me where I would add the parameter?

I didn't find in this project :/

**Comment by crookedneighbor** [comment]
My mistake the `vaultManager` option is a `boolean`, not an object: https://github.com/braintree/braintree-web-drop-in/blob/bf05e9a9ad10a7b22ccc80fa2e86561d16d7ed72/src/index.js#L269

Making it a top level object with the payment methods to hide as the keys should be fine.

**Comment by ale-grosselle** [comment]
@crookedneighbor Implemented the param if you want to take a look!

**Comment by crookedneighbor** [comment]
Made a couple of adjustments, added a changelog entry and rebased. I force pushed the changes, so if you have this checked out locally, you'll need to fetch the changes and reset to the remote branch if you need to make any changes.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # CHANGELOG
 
+## unreleased
+
+  - Add `vaultedPaymentMethodTypesThatShouldBeHidden` option to `dropin.create`
```
Comment (crookedneighbor) [inline CHANGELOG.md:5]: I'm not super happy with this name for the param, but I also can't think of a better name that describes what it does.

@braintree/team-sdk any of y'all have ideas for this?

**Comment by ale-grosselle** [comment]
Hello @crookedneighbor, any news?

**Review by shango420 — APPROVED** [review]

**Comment by shango420** [comment]
> 
Perfecto buen trabajo

**Comment by ale-grosselle** [comment]
Any news, could we merge it please? 🙏 🙏

**Comment by shango420** [comment]
> # **_~~~~_**

**Comment by ale-grosselle** [comment]
Hello @crookedneighbor, any news?

**Comment by armandodlvr** [comment]
@ale-grosselle  could you please rename `vaultedPaymentMethodTypesThatShouldBeHidden` to `hiddenVaultedPaymentMethodTypes` ? and resolve conflicts to merge your PR. Thanks

**Comment by ale-grosselle** [comment]
Hi @armandodlvr done!

File: src/index.js [hunk 2]
```diff
@@ -235,7 +235,11 @@ var VERSION = '__VERSION__';
...
- * @param {array} [options.paymentOptionPriority] Use this option to indicate the order in which enabled payment options should appear when multiple payment options are enabled. By default, payment options will appear in this order: `['card', 'paypal', 'paypalCredit', 'venmo', 'applePay', 'googlePay']`. Payment options omitted from this array will not be offered to the customer.
+ * @param {string[]} [options.paymentOptionPriority] Use this option to indicate the order in which enabled payment options should appear when multiple payment options are enabled. By default, payment options will appear in this order: `['card', 'paypal', 'paypalCredit', 'venmo', 'applePay', 'googlePay']`. Payment options omitted from this array will not be offered to the customer.
+ * @param {string[]} [options.hiddenVaultedPaymentMethodTypes] Use this option to hide certain types of vaulted payment methods from being displayed to the customer. Apple Pay, Google Pay and Venmo are always hidden (as those payment methods are only availabe as vaulted payment methods for existing subscriptions, new transactions or subscriptions must be newly authorized). Available options:
+ * * `'card'`
+ * * `'paypal'`
```
Comment (shango420) [inline src/index.js:241]: - [ ] xhangtechma@gmail.com ese es mi paypal

File: test/unit/dropin-model.js [hunk 3]
```diff
@@ -1591,5 +1591,72 @@ describe('DropinModel', () => {
...
     );
+
+    describe('hiddenVaultedPaymentMethodTypes options', function () {
+      test.each([
+        {
+          hiddenVaultedPaymentMethodTypes: [],
+          expected: ['AndroidPayCard', 'ApplePayCard', 'VenmoAccount']
+        },
+        {
+          hiddenVaultedPaymentMethodTypes: ['paypal'],
```
Comment (shango420) [inline test/unit/dropin-model.js:1602]: (xhangtechma@gmail.com)

**Review by shango420 — APPROVED** [review]
Mi paypal es (xhangtechma@gmail.com

**Comment by shango420** [comment]
> ### Summary
> 
> ~~Implemented `vaultedPaypalAccountsDisabled` argument.  
> If set to true, it force to disable paypal vault.~~
> 
> Implemented `vaultedPaymentMethodTypesThatShouldBeHidden`. 
> It accepts an array of paymentMethodTypes (e.g CreditCard, PayPalAccount) and specify any vaulted payment method type that want to block. Important: ApplePayCard, VenmoAccount, AndroidPayCard are always blocked.
> 
> This is very useful when you have two different payment flows and on one it is not possible to enable paypal vaulting (due to legal restriction for example).
> 
> ### Checklist
> 
> - [ ] Added a changelog entry
> 
> ### Authors
> > List GitHub usernames for everyone who contributed to this pull request.
> 
> - 
>

File: test/unit/dropin-model.js (hunk 3 above)
Comment (shango420) [reply test/unit/dropin-model.js:1602]: 4f0a48821db296c3ccdfbd25e613da5331e8d664

**Comment by armandodlvr** [comment]
@ale-grosselle the changelog was removed. could you add it?... thanks

**Comment by ale-grosselle** [comment]
> hiddenVaultedPaymentMethodTypes

Done!

> @ale-grosselle the changelog was removed. could you add it?... thanks

File: src/dropin-model.js [hunk 4]
```diff
@@ -461,8 +478,44 @@ function isPaymentOptionEnabled(paymentOption, options) {
...
+      ) {
+        // don't add the same payment method type a second time
+        return;
+      }
+
+      hiddenVaultedPaymentMethodTypes.push(paymentMethodId);
+    });
+  }
+
+  return hiddenVaultedPaymentMethodTypes;
```
Comment (shango420) [inline src/dropin-model.js:518]: https://paypal.me/xhangtech?country.x=MX&locale.x=es_XC

**Review by shango420 — APPROVED** [review]

**Comment by ale-grosselle** [comment]
@armandodlvr is it okay now?

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,9 @@
 # CHANGELOG
 
+## unreleased
+
+- Add `hiddenVaultedPaymentMethodTypes` option to `dropin.create`
```
Comment (cgdibble) [inline CHANGELOG.md:5]:
```suggestion
  - Add `hiddenVaultedPaymentMethodTypes` option to `dropin.create`
```

File: src/dropin-model.js [hunk 6]
```diff
@@ -13,7 +13,10 @@ var paymentSheetViews = require('./views/payment-sheet-views');
 var vaultManager = require('braintree-web/vault-manager');
 var paymentOptionsViewID = require('./views/payment-options-view').ID;
 
-var VAULTED_PAYMENT_METHOD_TYPES_THAT_SHOULD_BE_HIDDEN = [
+// these vaulted payment methods can only be used for existing subscription
+// any new transactions or subscriptons should prompt the customer to
+// authorize them again before using them
```
Comment (cgdibble) [inline src/dropin-model.js:18]:
```suggestion
// authorize them again before using them and thus should always be hidden.
```

File: test/unit/dropin-model.js [hunk 7]
```diff
@@ -1591,5 +1591,72 @@ describe('DropinModel', () => {
...
+            'VenmoAccount'
+          ]
+        },
+        {
+          // eslint-disable-next-line no-undefined
+          vaultedPaypalAccountsDisabled: undefined,
+          expected: ['AndroidPayCard', 'ApplePayCard', 'VenmoAccount']
+        }
+      ])(
+        'when hiddenVaultedPaymentMethodTypes is set to $vaultedPaypalAccountsDisabled',
```
Comment (cgdibble) [inline test/unit/dropin-model.js:1629]: Can you refine the test name to include some manner of expectation the test will evaluate? That is, when `hiddenVaultedPaymentMethodTypes` is set to <something>, what do we expect?

**Review by cgdibble — COMMENTED** [review]
@ale-grosselle just a couple last little things that are small nits, but otherwise it's looking excellent, thanks! Once those are in, we'll merge it.

File: src/dropin-model.js (hunk 6 above)
Comment (shango420) [reply src/dropin-model.js:18]: 🎉

File: test/unit/dropin-model.js (hunk 7 above)
Comment (shango420) [reply test/unit/dropin-model.js:1629]: > Can you refine the test name to include some manner of expectation the test will evaluate? That is, when `hiddenVaultedPaymentMethodTypes` is set to <something>, what do we expect?

Gracias por todo

Comment (cgdibble) [reply test/unit/dropin-model.js:1629]: @ale-grosselle if these last couple things get addressed then we can get this merged.

**Review by shango420 — APPROVED** [review]

File: src/index.js (hunk 2 above)
Comment (shango420) [reply src/index.js:241]: https://api.sandbox.paypal.com/v1/oauth2/token

Comment (shango420) [reply src/index.js:241]: This sentence uses `$` delimiters to show

**Comment by shango420** [comment]
vaultedPaymentMethodTypesThatShouldBeHidden

File: test/unit/dropin-model.js (hunk 7 above)
Comment (ale-grosselle) [reply test/unit/dropin-model.js:1629]: Hey @cgdibble thanks for the review! 
In these days I'm going to implement all

**Review by shango420 — APPROVED** [review]
39b8c62083b9d6809e540f2905d9994a9fd650ac

File: CHANGELOG.md (hunk 5 above)
Comment (shango420) [reply CHANGELOG.md]: > ```suggestion
>   - Add `hiddenVaultedPaymentMethodTypes` option to `dropin.create`
> ```

[]()}[@xhangtech/paypal.com}

Comment (shango420) [reply CHANGELOG.md]: > ```suggestion
>   - Add `hiddenVaultedPaymentMethodTypes` option to `dropin.create`
> ```

File: test/unit/dropin-model.js (hunk 7 above)
Comment (shango420) [reply test/unit/dropin-model.js:1629]: > Can you refine the test name to include some manner of expectation the test will evaluate? That is, when `hiddenVaultedPaymentMethodTypes` is set to _something_, what do we expect?

braintree.client.create({
  authorization: 'authorization'
}).then(function (clientInstance) {
  return braintree.paypalCheckout.create({
    client: clientInstance
  });
}).then(function (paypalCheckoutInstance) {
  return paypal.Button.render({
    env: 'production', // or 'sandbox'

    payment: function () {
      return paypalCheckoutInstance.createPayment({
        // your createPayment options here
      });
    },

    onAuthorize: function (data, actions) {
      // some logic here before tokenization happens below
      return paypalCheckoutInstance.tokenizePayment(data).then(function (payload) {
        // Submit payload.nonce to your server
      });
    }
  }, '#paypal-button');
}).catch(function (err) {
 console.error('Error!', err);
});

Comment (ale-grosselle) [reply test/unit/dropin-model.js]: @cgdibble done, could you check please?

**Comment by ale-grosselle** [comment]
Hello @cgdibble @crookedneighbor, 
is possible to merge this PR?

Thanks

**Review by cgdibble — APPROVED** [review]

**Comment by cgdibble** [comment]
I'm going to have some teammates take a look as an extra set of eyes in the morning, but otherwise it looks good and we should be able to merge it tomorrow.

**Review by bywood — APPROVED** [review]

**Review by ibooker — APPROVED** [review]

**Pull request merged**